src/fifo.ts

```diff
@@ -1,85 +1,93 @@
-// ported from https://www.npmjs.com/package/fast-fifo
-
-class FixedFIFO<T> {
-  public buffer: Array<T | undefined>
-  private readonly mask: number
-  private top: number
-  private btm: number
-  public next: FixedFIFO<T> | null
-
-  constructor (hwm: number) {
-    if (!(hwm > 0) || ((hwm - 1) & hwm) !== 0) {
-      throw new Error('Max size for a FixedFIFO should be a power of two')
-    }
+import defer, { DeferredPromise } from 'p-defer'
 
-    this.buffer = new Array(hwm)
-    this.mask = hwm - 1
-    this.top = 0
-    this.btm = 0
-    this.next = null
-  }
+export interface Next<T> {
+  done?: boolean
+  error?: Error
+  value?: T
+}
 
-  push (data: T) {
-    if (this.buffer[this.top] !== undefined) {
-      return false
-    }
+export type NextResult<T> = { done: false, value: T} | { done: true }
 
-    this.buffer[this.top] = data
-    this.top = (this.top + 1) & this.mask
+export interface FIFOOptions {
+  highWaterMark?: number
+  objectMode?: boolean
+}
 
-    return true
+export class FIFO<T> {
+  private readonly objectMode: boolean
+  private readonly highWaterMark: number
+  private queue: Array<Next<T>>
+  private deferred?: DeferredPromise<void>
+  private size: number
+
+  constructor (options: FIFOOptions = {}) {
+    this.objectMode = Boolean(options.objectMode)
+    this.highWaterMark = options.highWaterMark ?? Infinity
+    this.queue = []
+    this.size = 0
   }
 
-  shift () {
-    const last = this.buffer[this.btm]
-
-    if (last === undefined) {
-      return undefined
+  async push (val: Next<T>) {
+    if (this.size < this.highWaterMark) {
+      this.queue.push(val)
+
+      if (val.done === false) {
+        if (this.objectMode) {
+          this.size++
+        } else if (val.value instanceof Uint8Array) {
+          this.size += val.value.byteLength
+        } else {
+          throw new Error('objectMode was false but tried to push non-Uint8Array value')
+        }
+      }
+
+      return
     }
 
-    this.buffer[this.btm] = undefined
-    this.btm = (this.btm + 1) & this.mask
-    return last
-  }
+    // wait for a slot in the queue to become available
+    this.deferred = defer()
+    await this.deferred.promise
```

Comment:
I think this doesn't work if I push 2 values without awaiting - I think you need to accept it and buffer it...

Reply:
👍 should be fixed now